### PR TITLE
Fix: add missing dependencies

### PR DESCRIPTION
**Pull request by mdjermanovic**
This PR:

* Adds missing dependencies to `package.json`. The lines are just copied from ESLint's `package.json` so we could think about upgrading them later (some of them have new major versions available).
* Adds `"engines"` to `package.json` to match ESLint's. This was necessary to fix linting errors.
* Partially enables linting. Run `npm run lint`.
* Fixes linting errors, except for `tests/` and `conf/eslint-all.js`, which are added to `ignorePatterns`.
* Removes `package-lock`, I think we don't need it.
* Fixes a bug with missing `this.` in `config-validator.js`.

Notes:

* `eslint-all.js` is broken in this package, but I think it's unused.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -31,11 +31,10 @@ const naming = require("./shared/naming");
 module.exports = {
 
     Legacy: {
-        // config-array
         ConfigArray,
         createConfigArrayFactoryContext,
         CascadingConfigArrayFactory,
-        ConfigArray,
+        ConfigArrayFactory,
```
Comment (mdjermanovic) [inline lib/index.js:37]: This was a duplicate.

File: lib/shared/config-validator.js [hunk 2]
```diff
@@ -266,10 +265,10 @@ module.exports = class ConfigValidator {
      * @returns {void}
      */
     validateConfigSchema(config, source = null) {
-        const validateSchema = this.validateSchema || ajv.compile(configSchema);
+        validateSchema = this.validateSchema || ajv.compile(configSchema);
```
Comment (mdjermanovic) [inline lib/shared/config-validator.js:268]: This was a `no-shadow` warning, and also `no-unused-vars` for the upper `let` variable.

I believe this reverts to the original version, where the intention was to cache `validateSchema`.

File: lib/shared/config-validator.js [hunk 3]
```diff
@@ -266,10 +265,10 @@ module.exports = class ConfigValidator {
...
 
         if (!validateSchema(config)) {
-            throw new Error(`ESLint configuration in ${source} is invalid:\n${formatErrors(validateSchema.errors)}`);
+            throw new Error(`ESLint configuration in ${source} is invalid:\n${this.formatErrors(validateSchema.errors)}`);
```
Comment (mdjermanovic) [inline lib/shared/config-validator.js:271]: This was probably a bug.

File: package.json [hunk 4]
```diff
@@ -35,13 +36,24 @@
...
     "eslint-config-eslint": "^6.0.0",
+    "eslint-plugin-jsdoc": "^22.1.0",
+    "eslint-plugin-node": "^11.1.0",
     "eslint-release": "^3.1.0",
     "mocha": "^8.1.1"
   },
   "dependencies": {
     "ajv": "^6.12.4",
     "debug": "^4.1.1",
+    "espree": "^7.3.0",
```
Comment (mdjermanovic) [inline package.json:47]: This should be probably either always in sync, or resolved from ESLint instead.

Comment (kaicataldo) [reply package.json:47]: Resolving this from ESLint seems like a better way to go :+1:

File: package.json [hunk 5]
```diff
@@ -35,13 +36,24 @@
...
+    "globals": "^12.1.0",
+    "ignore": "^4.0.6",
     "import-fresh": "^3.2.1",
+    "js-yaml": "^3.13.1",
```
Comment (kaicataldo) [inline package.json:51]: Question: Does moving these dependencies over here allow us to remove any from ESLint?

**Comment by kaicataldo** [comment]
> eslint-all.js is broken in this package, but I think it's unused.

Can you elaborate on this?

**Comment by mdjermanovic** [comment]
> > eslint-all.js is broken in this package, but I think it's unused.
> 
> Can you elaborate on this?

Requires `require("../lib/rules")`, but rules don't exist in this package.

File: package.json (hunk 5 above)
Comment (mdjermanovic) [reply package.json:51]: Probably, we could try with deleting some and see if `node/no-extraneous-require` warns.

Comment (nzakas) [reply package.json:51]: I wouldn't remove dependencies from ESLint at this point. I'm still trying to get this package working correctly. :)

File: lib/index.js (hunk 1 above)
Comment (nzakas) [reply lib/index.js:37]: Thanks.

File: lib/shared/config-validator.js (hunk 2 above)
Comment (nzakas) [reply lib/shared/config-validator.js:268]: Looks like you're correct. Nice catch.

File: package.json (hunk 4 above)
Comment (nzakas) [reply package.json:47]: I'd say we can do this in the short-term, but in the long-term this should be passed in from ESLint. (Right now, this code isn't actually being used in ESLint, so I don't think it should cause issues either way.)

**Pull request merged**